packages/@aws-cdk/aws-elasticsearch/lib/domain.ts

```diff
@@ -0,0 +1,1503 @@
+  }
+
+  private grant(
+    grantee: iam.IGrantable,
+    domainActions: string[],
+    resourceArn: string,
+    ...otherResourceArns: string[]
+  ): iam.Grant {
+    const resourceArns = [resourceArn, ...otherResourceArns];
+
+    const grant = iam.Grant.addToPrincipal({
+      grantee,
+      actions: domainActions,
+      resourceArns,
+      scope: this,
+    });
+
+    return grant;
+  }
+}
+
+
+/**
+ * Reference to an Elasticsearch domain.
+ */
+export interface DomainAttributes {
+  /**
+   * The ARN of the Elasticsearch domain.
+   */
+  readonly domainArn: string;
+
+  /**
+   * The domain endpoint of the Elasticsearch domain.
+   */
+  readonly domainEndpoint: string;
+}
+
+
+/**
+ * Provides an Elasticsearch domain.
+ */
+export class Domain extends DomainBase implements IDomain {
+  /**
+   * Creates a Domain construct that represents an external domain via domain endpoint.
+   *
+   * @param scope The parent creating construct (usually `this`).
+   * @param id The construct's name.
+   * @param domainEndpoint The domain's endpoint.
+   */
+  public static fromDomainEndpoint(
+    scope: cdk.Construct,
+    id: string,
+    domainEndpoint: string,
+  ): IDomain {
+    const stack = cdk.Stack.of(scope);
+    const domainName = extractNameFromEndpoint(domainEndpoint);
+    const domainArn = stack.formatArn({
+      service: 'es',
+      resource: 'domain',
+      resourceName: domainName,
+    });
+
+    return Domain.fromDomainAttributes(scope, id, {
+      domainArn,
+      domainEndpoint,
+    });
+  }
+
+  /**
+   * Creates a Domain construct that represents an external domain.
+   *
+   * @param scope The parent creating construct (usually `this`).
+   * @param id The construct's name.
+   * @param attrs A `DomainAttributes` object.
+   */
+  public static fromDomainAttributes(scope: cdk.Construct, id: string, attrs: DomainAttributes): IDomain {
+    const { domainArn, domainEndpoint } = attrs;
+    const domainName = extractNameFromEndpoint(domainEndpoint);
+
+    return new class extends DomainBase {
+      public readonly domainArn = domainArn;
+      public readonly domainName = domainName;
+      public readonly domainEndpoint = domainEndpoint;
+
+      constructor() { super(scope, id); }
+    };
+  }
+
+  public readonly domainArn: string;
+  public readonly domainName: string;
+  public readonly domainEndpoint: string;
+  public readonly slowSearchLogGroup?: logs.ILogGroup;
+  public readonly slowIndexLogGroup?: logs.ILogGroup;
+  public readonly appLogGroup?: logs.ILogGroup;
+
+
+  private readonly domain: CfnDomain;
+
+  constructor(scope: cdk.Construct, id: string, props: DomainProps) {
+    super(scope, id, {
+      physicalName: props.domainName,
+    });
+
+    // If VPC options are supplied ensure that the number of subnets matches the number AZ
+    if (props.vpcOptions?.subnets.map((subnet) => subnet.availabilityZone).length != props?.zoneAwareness?.availabilityZoneCount) {
```

Comment:
Need to dedup.

```suggestion
    if (new Set(props.vpcOptions?.subnets.map((subnet) => subnet.availabilityZone)).size != props?.zoneAwareness?.availabilityZoneCount) {
```

Also - are we sure it has to exactly match the count? Intuitively seems like this should be `<`.

Reply:
Made both the Set and the `<` change.